tensorflow/core/kernels/conv_ops_gpu_3.cu.cc

```diff
@@ -413,6 +429,313 @@ struct PadInput<GPUDevice, T, int, NDIMS> {
+    // of the request and increase that parameter accordingly.
+    const bool long_side_request_not_satisfied =
+        max(tile_size_i, tile_size_j) > TileLongSide;
+
+    if (long_side_request_not_satisfied) {
+      BatchNarrowMatrixTransposeDispatcher<
+          T, TileLongSide * 2, TileShortSide>::DoIt(d, tile_size_i, tile_size_j,
+                                                    total_tiles_count, input,
+                                                    input_dims, output);
+    } else {
+      BatchNarrowMatrixTransposeDispatcher<
+          T, TileLongSide, TileShortSide + 1>::DoIt(d, tile_size_i, tile_size_j,
+                                                    total_tiles_count, input,
+                                                    input_dims, output);
+    }
+  }
+};
+
+template <typename T, int TileLongSide, int TileShortSide>
+struct BatchNarrowMatrixTransposeDispatcher<
+    T, TileLongSide, TileShortSide,
+    typename std::enable_if<TileSizeOnFrontier(TileLongSide, TileShortSide,
+                                               sizeof(T)),
+                            void>::type> {
+  static void DoIt(const GPUDevice& d, int tile_size_i, int tile_size_j,
+                   int total_tiles_count, const T* input,
+                   const Dimension<3>& input_dims, T* output) {
+    static_assert(
+        (TileLongSide & (TileLongSide - 1)) == 0,
+        "The length of the longer side of the tile is always a power of 2.");
+
+    constexpr int NumThreads = TileLongSide;
+    if (tile_size_i <= TileLongSide && tile_size_j <= TileShortSide) {
+      SwapDimension1And2InTensor3UsingTiles<T, NumThreads, TileLongSide,
+                                            TileShortSide>
+          <<<total_tiles_count, NumThreads, 0, d.stream()>>>(input, input_dims,
+                                                             output);
+    } else {
+      SwapDimension1And2InTensor3UsingTiles<T, NumThreads, TileShortSide,
+                                            TileLongSide>
+          <<<total_tiles_count, NumThreads, 0, d.stream()>>>(input, input_dims,
+                                                             output);
+    }
+  }
+};
+
+template <typename T, int TileLongSide, int TileShortSide>
+struct BatchNarrowMatrixTransposeDispatcher<
+    T, TileLongSide, TileShortSide,
+    typename std::enable_if<TileSizePastFrontier(TileLongSide, TileShortSide,
+                                                 sizeof(T)),
+                            void>::type> {
+  static void DoIt(const GPUDevice& d, int tile_size_i, int tile_size_j,
+                   int total_tiles_count, const T* input,
+                   const Dimension<3>& input_dims, T* output) {
+    assert(false &&
+           "BatchNarrowMatrixTransposeDispatcher has requested an unexpected "
+           "launch configuration. ");
+  }
+};
+
+// This function tries to recover, in a brute force way, the frontier defined in
```

Comment:
Nit, as written it sounds like this isn't guaranteed to succeed?

Reply:
Hi, this indeed has an unusually high chance of failure in the case when someone changes the tile spec and forgets to end the short side len definition with 2.

Although I did not consciously thought this way when writing this comment, I believe the wording did reflect my reservation.

Reply:
Ah, I see.  Do you think it would be worthwhile to say that, so as to help out someone who might hit this failure?  Maybe something like: "Builds the frontier defined in TileSizePossibilityFrontierCheck as a vector of tile size combinations.  This will fail if ...".